### PR TITLE
Monitor the Github Critical Workflows

### PR DESCRIPTION
**Pull request by prudhvigodithi**
### Description
Monitor the Github Critical Workflows

### Issues Resolved
Part of https://github.com/opensearch-project/opensearch-build/issues/4941

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by prudhvigodithi** [comment]
This PR https://github.com/opensearch-project/opensearch-metrics/pull/82 should be merged 1st to ensure that the resources has the right roles, alarms, metrics, namespaces created for GitHub app to publish the alarm metrics.
Thank you

File: README.md [hunk 1]
```diff
@@ -64,13 +64,14 @@ When you run the above command, the following takes place:
...
 | INSTALLATION_ID             | String  | ''        | Installation Id of your GitHub App, must install the App to repositories before retrieving the id.                                                                 | '1234567890'                              |
 | ADDITIONAL_RESOURCE_CONTEXT | Boolean | false     | Setting true will let each resource defined in RESOURCE_CONFIG to call GitHub Rest API and GraphQL for more detailed context (ex: node_id). Increase startup time. | true / false                              |
-| SERVICE_NAME                | String  | 'default' | Set Service Name                                                                                                                                                   | 'My Service'                              |'
+| SERVICE_NAME                | String  | 'default' | Set Service Name                                                                                                                                                   | 'My Service'                              | '   |
```
Comment (peterzhuamazon) [inline README.md:74]: I think you need to remove the extra `'` and format again.

File: src/call/github-workflow-runs-monitor.ts [hunk 2]
```diff
@@ -70,4 +81,35 @@ export default async function githubWorkflowRunsMonitor(app: Probot, context: an
   } catch (error) {
     app.log.error(`Error indexing log data: ${error}`);
   }
+
+  let count: any;
```
Comment (peterzhuamazon) [inline src/call/github-workflow-runs-monitor.ts:85]: What count? should be number or any here.

Comment (peterzhuamazon) [inline src/call/github-workflow-runs-monitor.ts:85]:
```suggestion
  let count: number;
```

File: src/call/github-workflow-runs-monitor.ts [hunk 3]
```diff
@@ -70,4 +81,35 @@ export default async function githubWorkflowRunsMonitor(app: Probot, context: an
...
+  if (job?.status === 'completed' && job?.conclusion === 'success') {
+    count = 0;
+  }
+  if (job?.status === 'completed' && job?.conclusion === 'failure') {
+    count = 1;
+  }
+  try {
+    const cloudWatchClient = new CloudWatchClient({ region: String(process.env.REGION) });
+    const putMetricDataCommand = new PutMetricDataCommand({
```
Comment (peterzhuamazon) [inline src/call/github-workflow-runs-monitor.ts:94]: Just wonder should we put these cloudwatch index in automation-app, or in metricscluster.

Comment (prudhvigodithi) [reply src/call/github-workflow-runs-monitor.ts:94]: This should be part of the automation app, github-workflow-runs-monitor.ts, resource Peter. We need to trigger a CloudWatch alarm immediately when a failure is detected. While most of the alarm setup is in the metrics cluster (and in the metrics repo: https://github.com/opensearch-project/opensearch-metrics/pull/82), we simply need to add a count to the alarm and index the data into the cluster. Monitoring workflows asynchronously after indexing to the metrics cluster would be too delayed. For instance, once data is added, we'd have to run a Lambda to query for failures (which is too late), or configure the alerting plugin to find the documents and alert us. However, the alerting plugin doesn’t easily capture passed metrics. Therefore, a straightforward solution would be to update the CloudWatch alarm immediately upon event detection.

File: src/call/github-workflow-runs-monitor.ts (hunk 2 above)
Comment (prudhvigodithi) [reply src/call/github-workflow-runs-monitor.ts:85]: For cloudwatch I see `Value: number | undefined`, if i change to number I get the error `Variable count is used before being assigned.`

File: README.md (hunk 1 above)
Comment (prudhvigodithi) [reply README.md:74]: These changes are the output of `npx prettier --write .`.

File: src/call/github-workflow-runs-monitor.ts (hunk 2 above)
Comment (prudhvigodithi) [reply src/call/github-workflow-runs-monitor.ts:85]: The count (0 or 1) that will be added to the cloudwatch alarm

File: README.md (hunk 1 above)
Comment (peterzhuamazon) [reply README.md:74]: I mean to say the original readme file has an extra `'` that should be removed which resulted in a wrong format.

File: src/call/github-workflow-runs-monitor.ts (hunk 2 above)
Comment (peterzhuamazon) [reply src/call/github-workflow-runs-monitor.ts:85]: > The count (0 or 1) that will be added to the cloudwatch alarm

Can we use a boolean instead?

Comment (peterzhuamazon) [reply src/call/github-workflow-runs-monitor.ts:85]: Then initialize the value to be 0 or 1 to start with.

Comment (prudhvigodithi) [reply src/call/github-workflow-runs-monitor.ts:85]: We cant use boolean as cloudwatch count value expects 0 or 1 peter.

Comment (prudhvigodithi) [reply src/call/github-workflow-runs-monitor.ts:85]: Let me check this.

File: src/call/github-workflow-runs-monitor.ts (hunk 3 above)
Comment (peterzhuamazon) [reply src/call/github-workflow-runs-monitor.ts:94]: I wonder if we should write a separate utility to handle cloudwatch, just like `utility/opensearch/opensearch-client`.

Comment (prudhvigodithi) [reply src/call/github-workflow-runs-monitor.ts:94]: Initially I thought that peter, but for now except for src/call/github-workflow-runs-monitor.ts I dont see any other use for now, in future we can always move to utility/opensearch/opensearch-client.

File: src/call/github-workflow-runs-monitor.ts (hunk 2 above)
Comment (prudhvigodithi) [reply src/call/github-workflow-runs-monitor.ts]: I have started with 0 and add if for `failure` or `startup_failure`, please check.

File: README.md (hunk 1 above)
Comment (prudhvigodithi) [reply README.md]: Updated can you check now?

File: src/call/github-workflow-runs-monitor.ts (hunk 3 above)
Comment (peterzhuamazon) [reply src/call/github-workflow-runs-monitor.ts:91]: I think we should always try to make these code re-usable.
Ex: Maintainers dashboards can face similar issue on this.

Comment (peterzhuamazon) [reply src/call/github-workflow-runs-monitor.ts:91]: Can you please start with an issue 1st.

Comment (prudhvigodithi) [reply src/call/github-workflow-runs-monitor.ts:91]: Ya then we can seperate into utility based on the use case peter.

**Review by peterzhuamazon — APPROVED** [review]

**Pull request merged**